text/0000-return-position-impl-trait-in-traits.md

```diff
@@ -0,0 +1,508 @@
+
+Consider the following trait:
+
+```rust
+trait IntoIntIterator {
+    fn into_int_iter(self) -> impl Iterator<Item = u32>;
+}
+```
+
+The semantics of this are analogous to introducing a new associated type within the surrounding trait;
+
+```rust
+trait IntoIntIterator { // desugared
+    type IntoIntIter: Iterator<Item = u32>;
+    fn into_int_iter(self) -> Self::IntoIntIter;
+}
+```
+
+When using `-> impl Trait`, however, there is no associated type that users can name.
+
+By default, the impl for a trait like `IntoIntIterator` must also use `impl Trait` in return position.
+
+```rust
+impl IntoIntIterator for Vec<u32> {
+    fn into_int_iter(self) -> impl Iterator<Item = u32> {
+        self.into_iter()
+    }
+}
+```
+
+It can, however, give a more specific type with `#[refine]`:[^refine]
+
+```rust
+impl IntoIntIterator for Vec<u32> {
+    #[refine]
+    fn into_int_iter(self) -> impl Iterator<Item = u32> + ExactSizeIterator {
+        self.into_iter()
+    }
+    
+    // ..or even..
+
+    #[refine]
+    fn into_int_iter(self) -> std::vec::IntoIter<u32> {
+        self.into_iter()
+    }
+}
+```
+
+Users of this impl are then able to rely on the refined return type, as long as the compiler can prove this impl specifically is being used. Conversely, in this example, code that is generic over the trait can only rely on the fact that the return type implements `Iterator<Item = u32>`.
+
+### async fn desugaring
+
+`async fn` always desugars to a regular function returning `-> impl Future`. When used in a trait, the `async fn` syntax can be used interchangeably with the equivalent desugaring in the trait and trait impl:
+
+```rust
+trait UsesAsyncFn {
+    // Equivalent to:
+    // fn do_something(&self) -> impl Future<Output = ()> + '_;
+    async fn do_something(&self);
+}
+
+// OK!
+impl UsesAsyncFn for MyType {
+    fn do_something(&self) -> impl Future<Output = ()> + '_ {
+        async {}
+    }
+}
+```
+```rust
+trait UsesDesugaredFn {
+    // Equivalent to:
+    // async fn do_something(&self);
+    fn do_something(&self) -> impl Future<Output = ()> + '_;
+}
+
+// Also OK!
+impl UsesDesugaredFn for MyType {
+    async fn do_something(&self) {}
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Equivalent desugaring for traits
+
+Each `-> impl Trait` notation appearing in a trait fn return type is effectively desugared to an anonymous associated type. In this RFC, we will use the placeholder name `$` when illustrating desugarings and the like.
+
+As a simple example, consider the following (more complex examples follow):
+
+```rust
+trait NewIntoIterator {
+    type Item;
+    fn into_iter(self) -> impl Iterator<Item = Self::Item>;
+}
+
+// becomes
+
+trait NewIntoIterator {
+    type Item;
+
+    type $: Iterator<Item = Self::Item>;
+
+    fn into_iter(self) -> <Self as NewIntoIterator>::$;
+}
+```
+
+```rust
+trait SomeTrait {
+    fn method<P0, ..., Pm>()
+}
+```
```

Comment:
What is this example meant to show?

Reply:
I think it was left by mistake, removing.